<img src="img/logo_negro.png" style="float: left; margin-top: 20px; width: 20%; ">
<h2 style="margin: 5px; margin-right: 0px; text-align: right">Dirección de Analítica</h2>
<h4 style="margin: 5px; margin-right: 0px; text-align: right">Marzo 1 de 2019</h4>
<h4 style="margin: 5px; margin-right: 0px; text-align: right">Creado por: <a href="mailto:jorge.quintana@datalytics.com">Jorge Quintana</a></h4>

<hr>
<h1 style="margin: 20px; margin-left: 0px; ">Demostración de Apache SuperSet en GCP usando Docker</h1>
<img src="img/super_gcp.png" style="margin: auto; height: 120px">
<p style = "text-align: justify; margin: 20px; margin-bottom: 20px; ">
Superset es una herramienta de visualización de datos desarrollado por AirBnB y donada a Apache Software Foundation y en estado de incubación bajo el patrocinio de Apache Incubator. Es definida como <b>"...a modern, enterprise-ready business intelligence web application..."</b>. Está desarrollada en Python, como Backend, y una mezcla de React.js, JavaScript y D3.js para el Frontend y utiliza el poder de los múltiples modulos con los que cuenta Python (SQLAlchemy, Pandas, Numpy, ...) para generar conexiones a diferentes fuentes de información -como lo son: Bases de Datos, Memorias Caches, DataLakes usando diferentes conectores- y procesansdo dichos datos por medio de una interfaz de SQL (SQLlab). Para más información véase <a href = "https://superset.incubator.apache.org">Apache Superset (Incubating)</a>
</p>
<hr>

### Caracterización de Victimas

Para esta prueba de concepto se importa un conjunto de datos disponoble en [Datos Abiertos](https://www.datos.gov.co/browse?q=Observatorio%20de%20paz&sortBy=relevance) sobre el registro y la caracterización de las victimas del conflicto armado y que han sido reportadas como beneficiarias de planes de ayuda económica. Los datos son reportados por el Ministerio de Interior y se encuentrar divididos en dos partes:

* Datos de Caracterización de Victimas
* Descripción de las Ayudas

Los datos son cargados mediante la API de Datos Abiertos. A primera vista el archivo de caracterización de las victimas tiene más registros que el archivo de descripción de las ayudas recibidas, por lo cual, se exploran para ver qué posible cruce se puede logar para mantener la mayor cantidad de registros con información completa. Para esta tarea, se define una función que genera la consulta para los datos solicitidados (parámetro 'dataset') y los une en una sola estructura de formato tabular, para cada uno de los dos archivos. Para la descripción de las victimas registradas se obtiene un Dataframe llamado **victimas** y para los datos de los beneficiarios de las ayudas económicas se obtiene un Dataframe llamado **beneficiarios**. Los resultados y el proceso de implementción de muestra a continuación.

In [1]:
import requests
import pandas as pd

"""
La función get_data genera las solicitudes de los datos a la API de Datos Abiertos mediante una petición GET. Recibe los siguientes argumentos de entrada:

    * dataset: Igual a 'victimas' si los datos a consultar son los registros de los victimas registrados o igual a 'beneficiarios' si la información
    requerida es la relacionada con los receptores de los auxilios económicos.
    * pags: Indica el número de páginas a consultar. Para el caso de victimas se tiene un total de 1900 registros, los cuales pueden ser consultados
    de diferentes formas, de acuerdo con la necesidad y capacidad de procesamiento de información mientras que los beneficiarios son poco menos de 800.
    * limit: Es la cantidad máxima de datos que se puede solicitar por página. Tiene un valor máximo de 1000 registros.
    
Adicionalmente, hay dos parámetro que son clave en la solicitud de datos a la API, pero que no son recibidos como parámetros de get_data():
    
    * offset: Indica el registro desde el cual de se debe hacer la siguiente búsqueda.
    * order: Ordena los resultados por id para que la paginación sea exitosa.
    
Finalmente, la función devuelve los dos Dataframes mencionados anteriormente.
"""

def get_data(dataset: str, pags: int, limit: int): 
    
    victimas = []
    beneficiarios = []
    if dataset == 'victimas':
        for i in range(pags):
            offset = i * limit
            aux = requests.get('https://www.datos.gov.co/resource/9kvn-3qq8.json?$order=:id&$limit={}&$offset={}'.format(limit, offset))
            if aux.text != '[]\n':
                aux = pd.read_json(aux.text)
                victimas.append(aux)
        victimas = pd.concat(victimas)
    else:
        for i in range(pags):
            offset = i * limit
            aux = requests.get('https://www.datos.gov.co/resource/g62i-exv7.json?$order=:id&$limit={}&$offset={}'.format(limit, offset))
            if aux.text != '[]\n':
                aux = pd.read_json(aux.text)
                beneficiarios.append(aux)
        beneficiarios = pd.concat(beneficiarios)
        
    return victimas, beneficiarios

Así las cosas, los resultados para los datos de caracterización de las victimas son los siguientes: (Sólo se muestra el encabezado con los 5 primeros registros, ya que en total son 1901 registros)

In [2]:
victimas, _ = get_data(dataset = 'victimas', pags = 2, limit = 1000)
print('Cantidad de registros para victimas: {}'.format(victimas.shape[0]))
victimas.head()

Cantidad de registros para victimas: 1901


actoterrorista adultomayor afrodescendiente amenaza apoyo  \
0             NO          NO               NO      NO    NO   
1             NO          NO               NO      NO    NO   
2             NO          NO               NO      NO    NO   
3             NO          NO               NO      NO    NO   
4             NO          NO               NO      NO    NO   

  articulacioninterinstitucional atencionpsicologica cabezadefamilia  \
0                             NO                  SI              SI   
1                             NO                  SI              NO   
2                             NO                  SI              SI   
3                             NO                  SI              NO   
4                             NO                  NO              NO   

      curso_de_vida            declaracion              ...                \
0           Adultez  DEFENSORIA DEL PUEBLO              ...                 
1           Adultez  DEFENSORIA DEL PUEBLO              ...                 
2           Jovenes  DEFENSORIA DEL PUEBLO              ...                 
3           Jovenes  DEFENSORIA DEL PUEBLO              ...                 
4  Primera infancia  DEFENSORIA DEL PUEBLO              ...                 

  orden orientacion personas_nucleo reclutamientoforzado rrom secuestro  \
0     1          NO               2                   NO   NO        NO   
1     2          NO               0                   NO   NO        NO   
2     3          NO               3                   NO   NO        NO   
3     4          NO               0                   NO   NO        NO   
4     5          NO               0                   NO   NO        NO   

        sexo tortura vigencia vinculaciondennagrujposarmados  
0   Femenino      NO     2017                             NO  
1  Masculino      NO     2017                             NO  
2   Femenino      NO     2017                             NO  
3  Masculino      NO     2017                             NO  
4  Masculino      NO     2017                             NO  

[5 rows x 37 columns]

Y la hacer lo mismo para obtener los datos de los auxilios económicos se obtienes los siguientes resultados:

In [3]:
_, beneficiarios = get_data(dataset = 'beneficiarios', pags = 2, limit = 1000)
print('Cantidad de registros para victimas receptoras de ayudas económicas: {}'.format(beneficiarios.shape[0]))
beneficiarios.head()

Cantidad de registros para victimas receptoras de ayudas económicas: 782


auxilio_econ_mico_mes curso_de_vida edad grupo_etario hecho_victimizante  \
0            02 FEBRERO       Adultez   47      29 a 59     DESPLAZAMIENTO   
1            02 FEBRERO       Jovenes   19      18 a 28     DESPLAZAMIENTO   
2            02 FEBRERO       Jovenes   25      18 a 28     DESPLAZAMIENTO   
3            02 FEBRERO       Adultez   38      29 a 59     DESPLAZAMIENTO   
4            02 FEBRERO       Adultez   49      29 a 59     DESPLAZAMIENTO   

   monto_asignado n_cleo_familiar  orden personas             sexo  vigencia  
0          687375          TIPO A      1        2         Femenino      2017  
1          687375          TIPO B      2        3         Femenino      2017  
2          687375          TIPO B      3        4         Femenino      2017  
3          687375          TIPO B      4        3  SIN INFORMACION      2017  
4          423000          TIPO A      5        1        Masculino      2017

Una conclusión es evidente, la cantidad de victimas registradas en el archivo de caracterización es mayor a la cantidad de victimas receptoras de auxilios económicos. Por otro lado, según la documentación de los datos exite un campo que permite el cruce de los conjuntos de datos, a saber: Orden. Este campo es un identificador númerico para cada una de las victimas. Así las cosas, el cruce de los datos se hace usando **orden** como llave. Los resultados se muestran a continuación.

In [4]:
data = pd.merge(victimas, beneficiarios, how = 'left', on = 'orden', indicator = True)
data.head()

actoterrorista adultomayor afrodescendiente amenaza apoyo  \
0             NO          NO               NO      NO    NO   
1             NO          NO               NO      NO    NO   
2             NO          NO               NO      NO    NO   
3             NO          NO               NO      NO    NO   
4             NO          NO               NO      NO    NO   

  articulacioninterinstitucional atencionpsicologica cabezadefamilia  \
0                             NO                  SI              SI   
1                             NO                  SI              NO   
2                             NO                  SI              SI   
3                             NO                  SI              NO   
4                             NO                  NO              NO   

    curso_de_vida_x            declaracion  ...   curso_de_vida_y edad_y  \
0           Adultez  DEFENSORIA DEL PUEBLO  ...           Adultez     47   
1           Adultez  DEFENSORIA DEL PUEBLO  ...           Jovenes     19   
2           Jovenes  DEFENSORIA DEL PUEBLO  ...           Jovenes     25   
3           Jovenes  DEFENSORIA DEL PUEBLO  ...           Adultez     38   
4  Primera infancia  DEFENSORIA DEL PUEBLO  ...           Adultez     49   

  grupo_etario_y hecho_victimizante monto_asignado n_cleo_familiar personas  \
0        29 a 59     DESPLAZAMIENTO       687375.0          TIPO A        2   
1        18 a 28     DESPLAZAMIENTO       687375.0          TIPO B        3   
2        18 a 28     DESPLAZAMIENTO       687375.0          TIPO B        4   
3        29 a 59     DESPLAZAMIENTO       687375.0          TIPO B        3   
4        29 a 59     DESPLAZAMIENTO       423000.0          TIPO A        1   

            sexo_y vigencia_y _merge  
0         Femenino     2017.0   both  
1         Femenino     2017.0   both  
2         Femenino     2017.0   both  
3  SIN INFORMACION     2017.0   both  
4        Masculino     2017.0   both  

[5 rows x 48 columns]

Como validación de la calidad de los datos luego del cruce se genera un conteo del campo **\_merge**, el cual identifica en qué base se encuentra el dato resultante, es decir, si el registro está presente en ambas bases **'both'**, sólo en la base de la izquierda **left_only** o sólo en la base de la derecha **right_only**. Los resultados muestran que, del total de 1901 registros de la base de registro de victimas sólo 782 cruzan con la base de beneficiarios, quedando un total de 1119 victimas sin infromación de los auxilios económicos recibidos.

In [5]:
data.groupby('_merge').count().vigencia_x

_merge
left_only     1119
right_only       0
both           782
Name: vigencia_x, dtype: int64

Como paso final, y antes de cargar los datos a PostgreSQL, se revisan los nombres de los campos en los datos resultantes y se eliminan las columnas repetidas, que en este caso corresponden a las columnas agregadas de la base de beneficiarios, y se renombran las que pertenecen a la base de victimas, pero removiendo la terminación de **'\_x'** que se agrega por defecto. Finalmente, se guardan los resultados en un archivo CSV.

In [7]:
remover = list(data.filter(like = '_y').columns)
modificar = list(data.filter(like = '_x').columns)
reemplazar_con = ['curso_de_vida', 'edad', 'grupo_etario', 'sexo', 'vigencia']
data.drop(remover, axis = 1, inplace = True)

for i in range(len(modificar)):
    data.rename(columns = {modificar[i] : reemplazar_con[i]}, inplace = True)
    
data.to_csv('data/data_final.csv', index = False)